pkg/daemon/update.go

```diff
@@ -2015,103 +2024,103 @@ func (dn *Daemon) reboot(rationale string) error {
 	if currentImage == desiredImage {
-		// Orrrr....if we've live updated to it
 		glog.Infof("Node is on proper image %s", desiredImage)
 
-	} else if liveUpdatedEquivalentTo == desiredImage {
-		glog.Info("No need to update, live update is equivalent")
 	} else {
+		// At this point we know we need to rebase
+		if err := dn.setWorking(); err != nil {
+			return fmt.Errorf("failed to set working: %w", err)
+		}
 
+		pullSecret, err := dn.GetPullSecret()
+		if err != nil {
+			return err
+		}
+
+		// TODO drop NodeUpdaterClient or change its interface
 		client := &RpmOstreeClient{}
+		if err := client.RebaseLayered(desiredImage, pullSecret); err != nil {
+			return err
+		}
+
 		state, err := client.GetState()
 		if err != nil {
 			return err
 		}
 
-		// Look through our deployments
-		// If we've rebased to it, but not booted, that's okay IF we've live-applied
-		// but we need to know whether we're live applying when we rebase
-		// ughhhh it really should be transactional
-
+		var staged Deployment
+		var activeChecksum string
 		for _, deployment := range state.Deployments {
-			// What we're looking for is at least in the list
-			if strings.TrimPrefix(deployment.ContainerImageReference, "ostree-unverified-registry:") == desiredImage {
-				// We rebased but we haven't booted, might be a liveapply
-				if deployment.Staged == true {
-					//TODO(jkyros): Check to see about liveapply
-					glog.Infof("Node is staged to %s, checking to see if we've liveapplied", desiredImage)
-					return nil
-				}
-
-				// Everything is perfect, we're already there the good way
-				if deployment.Booted == true {
-					glog.Infof("Node is already in image %s", desiredImage)
-					//TODO(jkyros): Add an annotation
-					return nil
+			if deployment.Staged {
+				staged = deployment
+			}
+			if deployment.Booted {
+				if deployment.LiveReplaced != "" {
+					activeChecksum = deployment.LiveReplaced
+				} else {
+					activeChecksum = deployment.Checksum
 				}
 			}
 		}
 
-		pullSecret, err := dn.GetPullSecret()
+		diffFileSet, err := Diff(activeChecksum, staged.Checksum)
 		if err != nil {
-			return err
+			return fmt.Errorf("failed to diff container images: %w", err)
 		}
-
-		//TODO(jkyros): what should these perms be
-		os.Mkdir("/run/ostree", 0544)
-
-		err = ioutil.WriteFile("/run/ostree/auth.json", pullSecret, 0400)
+		actions, err := dn.calculatePostConfigChangeActionFromFiles(diffFileSet)
 		if err != nil {
 			return err
 		}
 
-		_, err = client.RebaseLayered(desiredImage)
-		if err != nil {
+		// Check and perform node drain if required
+		readOldFile := func(path string) ([]byte, error) {
+			return Cat(currentImage, path)
+		}
+		readNewFile := func(path string) ([]byte, error) {
+			return Cat(desiredImage, path)
+		}
+		if err := dn.drainIfRequired(actions, diffFileSet, readOldFile, readNewFile); err != nil {
 			return err
 		}
 
-		//defer os.Unlink("/run/ostree/auth.json")
```

Comment:
Do we want to leave it there if we apply live? Agreed probably doesn't matter much

Reply:
I left it there in case I wanted to horse with it manually -- I figured if I had to "drive it out of the ditch" because I broke it, I didn't want to have to go get credentials too :)

Reply:
Is there any risk to leaving credentials lying around? If not I'll take out the removal I put back in

Reply:
I don't think there's a real risk here, *but* we should be sure the file ownership/permissions are `root:root` and `0600`.